Student models & tests for TRILL distillation ICASSP submission

jpeplins opened the pull request:
- Added "MobileNetV3Tiny" Implementation
- Added option to create student model without frontend (In support of distilling on pre-computed spectrograms and targets).
- Added option to average-pool MobileNet output.
- Added option to exclude bottleneck layer.
- Unit tests for above functionality.

jpeplins commented:
@googlebot I signed it!

joel-shor commented on non_semantic_speech_benchmark/distillation/models.py:
```diff
@@ -17,42 +17,92 @@
-  embeddings = tf.keras.backend.batch_flatten(model_out)
-  embeddings.set_shape([None, 3 * 2 * 1280])
+  if avg_pool:
+    model_out = model(feats)
```
This line is common, so you can pull this out of the `if` block

jpeplins replied:
Done.

joel-shor commented on non_semantic_speech_benchmark/distillation/models.py:
```diff
@@ -17,42 +17,92 @@
-      bottleneck_dimension, name='distilled_output')(embeddings)
-  output = tf.keras.layers.Dense(
-      output_dimension, name='embedding_to_target')(embeddings)
+  output = tf.keras.layers.Dense(output_dimension, activation=tf.nn.swish, name='embedding_to_target')(embeddings)
```
Why did you add `tf.nn.swish` to the final layer? that seems...wrong, no?

jpeplins replied:
Good catch - see my below comment.

joel-shor commented on non_semantic_speech_benchmark/distillation/models.py:
```diff
@@ -17,42 +17,92 @@
+  if bottleneck_dimension:
+    embeddings = tf.keras.layers.Flatten()(model_out)
+    embeddings = tf.keras.layers.Dense(
+      bottleneck_dimension, activation=tf.nn.swish, name='distilled_output')(embeddings)
```
same here. why?

jpeplins replied:
This was committed in error. 
Empirically it seems to speed up training (with no observable difference in student quality), but without intuition to justify it.

joel-shor reviewed (approved):
Nice

joel-shor commented:
Do you know why there's a travis error?

jpeplins commented:
Travis errors are build issues outside of our project. Seems unrelated to these changes.

srjoglekar246 commented:
Hey Jacob, I see multiple PyLint errors internally, which is blocking submission (for example, 'Access to a protected member _inverted_res_block of a client class' & `lambda x: _parse_examples(x)` -> `_parse_examples`). Can you resolve these (I think yapf might be useful)?

jpeplins commented:
Hey Sachin, two things:

1. I added a few extra flags to control the learning rate schedule. This is unrelated to the linting issues above.

2. I seemed to have fixed all of the problematic formatting issues. My linter picked up a lot of things that were already committed to the codebase, so I'm assuming those particular issues aren't blocking. To that end, I fixed every _new_ formatting issue my code introduced, with the exception of `Access to a protected member _inverted_res_block`, where I opted to disable the check (using `pylint:disable=protected-access`). This saves me from duplicating a lot of mobilenet code. 

Let me know if you run into any issues.

srjoglekar246 commented:
I submitted Jacob's code with lint changes in [this commit](https://github.com/google-research/google-research/commit/b45380774d8258bceeb5175291eb1e80a3664c3d).

joel-shor commented:
Excellent! Thanks, Sachin and Jacob.

On Fri, Nov 20, 2020, 3:24 AM Sachin Joglekar <notifications@github.com>
wrote:

> I submitted Jacob's code with lint changes in this commit
> <https://github.com/google-research/google-research/commit/b45380774d8258bceeb5175291eb1e80a3664c3d>
> .
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/google-research/google-research/pull/452#issuecomment-730554967>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABN567F5OOWYEL6I3MSTHT3SQVPHFANCNFSM4TKOQ5CA>
> .
>

jpeplins commented:
Thanks Sachin! I'll close this PR.
